### ISSUE TITLE
[$250] Tag - No error when creating new tag with existing name when tag name has X: Y format

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#45067]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.5-5
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4704925
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to Workspace settings > Tag
3. Add a tag in X: Y format
4. Click Add tag
5. Attempt to add the same tag in Step 3
6. Note that there is no error that the same tag already exists
7. Create a different tag
8. Click on the tag in Step 7
9. Rename it to the same tag in Step 3 and save it
10. Note that there is no error too and the old tag in Step 7 is gone

## Expected Result:
When the tag name has X: Y format, there will be error indicating that the same tag already exists when creating a new tag with the same name (Step 5) and renaming tag to the existing tag name (Step 10)

## Actual Result:
When the tag name has X: Y format, no error when creating a new tag with the same name (Step 5) and renaming tag to the same name (Step 10)

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/355b6472-f6b5-476a-890f-44205561a325

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0126cdffc6828dd304</li>
        <li>Upwork Job ID: 1811543862318164445</li>
        <li>Last Price Increase: 2024-07-25</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @getusha</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave-collect - Release 1

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There is no tag existing error when trying to add or edit a tag in X: Y format.

### What is the root cause of that problem?
When we add a tag in X: Y format, the colon (:) is escaped, so it becomes `X\: Y`. When we check whether there is an existing tag, we just use the tag name input from the user without escaping it.
https://github.com/Expensify/App/blob/9c32293ab6c0465eec38f1c573b8b5a09e1e112d/src/pages/workspace/tags/WorkspaceCreateTagPage.tsx#L48-L49

For example, if we input `a: b`, the tag will be saved as `a\: b`. When we try to add `a: b` again, it will check if `a: b` exists or not which doesn't.

### What changes do you think we should make in order to solve the problem?

Escape the tag name before doing the comparison.
https://github.com/Expensify/App/blob/9c32293ab6c0465eec38f1c573b8b5a09e1e112d/src/pages/workspace/tags/WorkspaceCreateTagPage.tsx#L43-L49

`PolicyUtils.escapeTagName(values.tagName.trim())`

We need to do the same for edit tag page.

**Comment by mallenexpensify** [comment]
Def seems like it can be External.  @getusha , please review the proposals above. 
Also, I'd love to fix any other similar _escaped_ bugs if we're able to do so at the same time

**Comment by getusha** [comment]
this fell through the cracks, will give an update today!

**Comment by getusha** [comment]
@bernhardoj the `escapeTagName` converts `x:y` -> `x\\:y` which will not be available in `tags` as a key, but i see the `name` value holds that value. and was able to get it work by changing the condition to `Object.values(tags).find(v => v.name === tagName)` could you check this and update your proposal?

@devguest07 Instead of converting the entire array to match the tag name value, I think the best option is to just change the tag name value to match the tag name format in the tags array.

**Comment by bernhardoj** [comment]
@getusha the solution works fine for me.

https://github.com/user-attachments/assets/e251349c-8183-4972-8ae3-3b675bda5632

> the escapeTagName converts x:y -> x\\:y which will not be available in tags as a key,

This is a similar concern that I answered [here](https://github.com/Expensify/App/pull/42381#discussion_r1606595753)

Do you have any tag name that doesn't work for you?

**Comment by getusha** [comment]
@bernhardoj could you try with `x\:y`?

**Comment by bernhardoj** [comment]
@getusha it works fine. If I add `x\:y` twice, I will get an error. I guess the issue you mean is that if we add `x:y` and `x\:y`, there will be 2 `x:y` in the tag list, is that correct?
<img width="828" alt="image" src="https://github.com/user-attachments/assets/6093c61a-61d0-4c89-8e77-118d89decdd3">

If yes, then it's the issue with how we display the tag, specifically the `getCleanedTagName`.
https://github.com/Expensify/App/blob/fa738c641e7d2b289850da7b4be11369c33b96d6/src/pages/workspace/tags/WorkspaceTagsPage.tsx#L108

If you remove `getCleanedTagName`, you will see both tags are different.
<img width="835" alt="image" src="https://github.com/user-attachments/assets/8bc13f8f-1ef1-4dab-8085-53e5d30fffbc">

We need to update `getCleanedTagName` to only remove 1 `\` (`tag?.replace(/\\:/g, CONST.COLON);`).
https://github.com/Expensify/App/blob/fa738c641e7d2b289850da7b4be11369c33b96d6/src/libs/PolicyUtils.ts#L289-L291

**Comment by mallenexpensify** [comment]
@getusha 👀 on @bernhardoj 's comment above plz

**Comment by getusha** [comment]
>it works fine. If I add x\:y twice, I will get an error. I guess the issue you mean is that if we add x:y and x\:y, there will be 2 x:y in the tag list, is that correct?

Yes, could you update your proposal to include that case? thanks

**Comment by bernhardoj** [comment]
[Proposal](https://github.com/Expensify/App/issues/45067#issuecomment-2217636309) Updated

**Comment by getusha** [comment]
@bernhardoj's [proposal](https://github.com/Expensify/App/issues/45067#issuecomment-2217636309) looks good to me.
🎀 👀 🎀  C+ Reviewed!

**Comment by getusha** [comment]
Friendly bump @techievivek

**Comment by bernhardoj** [comment]
PR is ready

cc: @getusha

**Comment by mallenexpensify** [comment]
@getusha _owes_ $500 from https://github.com/Expensify/App/issues/32699   , let's use this issue to knock it down by $250, then we'll find another $250 one.

**Comment by bernhardoj** [comment]
I opened a new PR to handle the edit tag case.

cc: @getusha @techievivek

**Comment by mallenexpensify** [comment]
@bernhardoj, @getusha  can you provide and update on where we're at here? I think payment is due now cuz the[ second PR hit production over a week ago. ](https://github.com/Expensify/App/pull/47289#issuecomment-2294143418)

And, with/for[ the first PR](https://github.com/Expensify/App/pull/46556#issuecomment-2294845576), I see 
> We've been saving tag name by escaping it before this PR, so this PR didn't cause it.

So that doesn't seems like it was a regression.

**Comment by mallenexpensify** [comment]
Contributor: @bernhardoj due $250 via NewDot
Contributor+: @getusha due $250 via NewDot 

@getusha we want a regression test here, right?  Assuming so, plz complete the BZ checklist below. Thx

**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@] Determine if we should create a regression test for this bug.
- [ ] [@] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@BZ] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:
- [ ]

**Comment by techievivek** [comment]
Gentle bump @getusha on above.

**Comment by techievivek** [comment]
Not overdue.

**Comment by getusha** [comment]
[@] The PR that introduced the bug has been identified. Link to the PR: **I wouldn't call this a regression at it seems to be an edge case that must've been missed on feature implementation.**
 [@] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/a**
 [@] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/a**
 [@] Determine if we should create a regression test for this bug. **Yes**
 [@] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Regression Test Proposal**

1. Open the app 
2. Go to Workspace settings > Tag
3. Add `X:Y` tag
4. Attempt to add the same tag `X:Y`
5. Verify `A tag with this name already exists` error appears
6. Add a different tag
7. Click on the tag in Step 6
8. Rename it to the same tag in Step 3 (`X:Y`) and save it
9. Verify `A tag with this name already exists` error appears

Do we agree 👍 or 👎

**Comment by techievivek** [comment]
Looks good, @mallenexpensify let's get it added to testRail, thanks.

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj

**Comment by mallenexpensify** [comment]
We good here! Test case created
- https://github.com/Expensify/Expensify/issues/424794
